src/database/database-sqlite3.cpp

```diff
@@ -191,40 +224,70 @@ MapDatabaseSQLite3::MapDatabaseSQLite3(const std::string &savedir):
 
 MapDatabaseSQLite3::~MapDatabaseSQLite3()
 {
-	FINALIZE_STATEMENT(m_stmt_read)
-	FINALIZE_STATEMENT(m_stmt_write)
-	FINALIZE_STATEMENT(m_stmt_list)
-	FINALIZE_STATEMENT(m_stmt_delete)
+	FINALIZE_STATEMENT(read)
+	FINALIZE_STATEMENT(write)
+	FINALIZE_STATEMENT(list)
+	FINALIZE_STATEMENT(delete)
 }
 
 
 void MapDatabaseSQLite3::createDatabase()
 {
-	assert(m_database); // Pre-condition
+	assert(m_database);
 
-	SQLOK(sqlite3_exec(m_database,
+	// Note: before 5.12.0 the format was blocks(pos INT, data BLOB).
+	// This function only runs for newly created databases.
+
+	const char *schema =
 		"CREATE TABLE IF NOT EXISTS `blocks` (\n"
-			"	`pos` INT PRIMARY KEY,\n"
-			"	`data` BLOB\n"
-			");\n",
-		NULL, NULL, NULL),
+			"`x` INT,"
+			"`y` INT,"
+			"`z` INT,"
+			"`data` BLOB,"
+			// Declaring a primary key automatically creates an index and the order
+			// decides which range operations can be sped up. Those are:
+			// * x BETWEEN ? AND ?
+			// * x = ? AND z BETWEEN ? AND ?
+			// * x = ? AND z = ? AND y BETWEEN ? AND ?
```

Comment:
It is slightly better than that since sqlite does skip scans: https://www.sqlite.org/optoverview.html#skipscan

Reply:
thanks for researching this. I just trusted @appgurueu since it matched my intuition.